streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -374,50 +379,68 @@ void handleLostAll() {
-        // A client could contain some inactive tasks whose states are still kept on the local storage in the following scenarios:
-        // 1) the client is actively maintaining standby tasks by maintaining their states from the change log.
-        // 2) the client has just got some tasks migrated out of itself to other clients while these task states
-        //    have not been cleaned up yet (this can happen in a rolling bounce upgrade, for example).
-
-        final Set<TaskId> locallyStoredTasks = new HashSet<>();
-
         final File[] stateDirs = stateDirectory.listTaskDirectories();
         if (stateDirs != null) {
             for (final File dir : stateDirs) {
                 try {
                     final TaskId id = TaskId.parse(dir.getName());
-                    // if the checkpoint file exists, the state is valid.
-                    if (new File(dir, StateManagerUtil.CHECKPOINT_FILE_NAME).exists()) {
-                        locallyStoredTasks.add(id);
+                    final Task task = tasks.get(id);
+                    if (task != null) {
+                        if (task.isActive() && task.state() == RUNNING) {
+                            taskOffsetSums.put(id, Task.LATEST_OFFSET);
+                        } else {
+                            taskOffsetSums.put(id, computeOffsetSum(task.changelogOffsets()));
+                        }
+                    } else {
+                        try {
+                            // if we are able to lock this task dir and find a valid checkpoint file, we are
+                            // responsible for encoding its offsets in our subscription
+                            final File checkpointFile = new File(dir, StateManagerUtil.CHECKPOINT_FILE_NAME);
+                            if (stateDirectory.lock(id) && checkpointFile.exists()) {
+                                taskOffsetSums.put(id, computeOffsetSum(new OffsetCheckpoint(checkpointFile).read()));
+                                lockedUnassignedTaskDirectories.add(id);
+                            }
+                        } catch (final IOException e) {
+                            // if for any reason we can't lock this task dir and read its checkpoint file, just move on
+                        }
                     }
                 } catch (final TaskIdFormatException e) {
-                    // there may be some unknown files that sits in the same directory,
-                    // we should ignore these files instead trying to delete them as well
+                    // we should just ignore any unknown files that sit in the same directory
                 }
             }
         }
 
-        return locallyStoredTasks;
+        return taskOffsetSums;
+    }
+
+    private long computeOffsetSum(final Map<TopicPartition, Long> changelogOffsets) {
+        long offsetSum = 0;
+        for (final long offset : changelogOffsets.values()) {
+            offsetSum += offset;
```

Comment:
what should we do if the offset is negative?

Reply:
Oh, also, should we detect overflow and pin to `MAX_VALUE` in that case?

Reply:
Seems like the only negative offset we can get is `-1`, which indicates the offset is unknown in which case we should skip it. Will add a check for overflow too

Reply:
Hmm. Skipping would count the "current position" on that store as `0`. Should we assume "unknown offset" equates to "fully caught up", or is it safer to set it to `MAX_VALUE`, or something else?

Reply:
-1 means that `recordMetadata.offset == ProduceResponse.INVALID_OFFSET`, which sounds to me like either the topic or producer isn't initialized yet or is corrupted in some way. Both of those make sense (to me) to interpret as an offset of `0`. But, I'm no expert in the Producer client and various offset meanings.

w.r.t the matter of overflow, I think we should aim to keep things simple and just pin to `MAX_VALUE` in the event of overflow, if we think that's likely to be a rare event. Obviously we should at the least make sure we don't crash or seriously harm the operation or results of the app -- pinning to `MAX_VALUE` will at worst make us potentially switch over to an active task before it's completely caught up.